rasa/nlu/featurizers/featurizer.py

```diff
@@ -1,48 +1,133 @@
+from rasa.shared.exceptions import InvalidConfigException
+from rasa.shared.nlu.constants import FEATURE_TYPE_SENTENCE, FEATURE_TYPE_SEQUENCE
 
+# TODO: remove after all featurizers have been migrated
+from rasa.nlu.featurizers._featurizer import (
+    Featurizer,
+    SparseFeaturizer,
+    DenseFeaturizer,
+)
 
-class Featurizer(Component):
-    def __init__(self, component_config: Optional[Dict[Text, Any]] = None) -> None:
-        if not component_config:
-            component_config = {}
+Featurizer = Featurizer
+SparseFeaturizer = SparseFeaturizer
+DenseFeaturizer = DenseFeaturizer
 
-        # makes sure the alias name is set
-        # Necessary for `unique_name` to be defined
-        self.component_config = component_config
-        component_config.setdefault(FEATURIZER_CLASS_ALIAS, self.unique_name)
+FeatureType = TypeVar("FeatureType")
 
-        super().__init__(component_config)
 
+class Featurizer2(Generic[FeatureType], ABC):
+    """Base class for all featurizers."""
 
-class DenseFeaturizer(Featurizer):
     @staticmethod
-    def _calculate_sentence_features(
-        features: np.ndarray, pooling_operation: Text
-    ) -> np.ndarray:
-        # take only non zeros feature vectors into account
-        non_zero_features = np.array([f for f in features if f.any()])
-
-        # if features are all zero just return a vector with all zeros
-        if non_zero_features.size == 0:
-            return np.zeros([1, features.shape[-1]])
-
-        if pooling_operation == MEAN_POOLING:
-            return np.mean(non_zero_features, axis=0, keepdims=True)
-
-        if pooling_operation == MAX_POOLING:
-            return np.max(non_zero_features, axis=0, keepdims=True)
-
-        raise ValueError(
-            f"Invalid pooling operation specified. Available operations are "
-            f"'{MEAN_POOLING}' or '{MAX_POOLING}', but provided value is "
-            f"'{pooling_operation}'."
-        )
+    def get_default_config() -> Dict[Text, Any]:
+        """Returns the component's default config."""
+        return {FEATURIZER_CLASS_ALIAS: None}
+
+    def __init__(self, name: Text, config: Dict[Text, Any]) -> None:
+        """Instantiates a new featurizer.
+
+        Args:
+          config: configuration
+          name: a name that can be used as identifier, in case the configuration does
+            not specify an `alias` (or this `alias` is None)
+        """
+        super().__init__()
+        self.validate_config(config)
+        self._config = {**self.get_default_config(), **config}
+        self._identifier = self._config[FEATURIZER_CLASS_ALIAS] or name
+
+    @classmethod
+    @abstractmethod
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the component is configured properly."""
+        ...
+
+    @classmethod
+    @abstractmethod
+    def validate_compatibility_with_tokenizer(cls, tokenizer: Tokenizer) -> None:
```

Comment:
tokenizer or `Type[Tokenizer]`? I'd actually prefer the later

Reply:
is now fixed to `validate_compatibility_with_tokenizer(config, tokenizer_type)` (and added a test I forgot 😬 )